test/e2e/node_drain_timeout.go

```diff
@@ -0,0 +1,168 @@
+		Expect(err).To(BeNil(), "Need a workload client to interact to the workload cluster")
+		framework.DeployUnevictablePod(ctx, framework.DeployUnevictablePodInput{
+			WorkloadCluster:                    workloadClient,
+			MachineDeployments:                 machineDeployments,
+			WaitForDeploymentAvailableInterval: input.E2EConfig.GetIntervals(specName, "wait-deployment-available"),
+		})
+
+		By("Scale the machinedeployment down to zero. If we didn't have the NodeDrainTimeout duration, the node drain process would block this operator.")
+		nodeDrainTimeoutMachineDeploymentInterval := convertMachineDeploymentDurationToInterval(machineDeployments[0].Spec.Template.Spec.NodeDrainTimeout)
+		for _, md := range machineDeployments {
+			framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+				ClusterProxy:              input.BootstrapClusterProxy,
+				Cluster:                   cluster,
+				MachineDeployment:         md,
+				WaitForMachineDeployments: nodeDrainTimeoutMachineDeploymentInterval,
+				Replicas:                  0,
+			})
+		}
+
+		By("Deploy workload on the master node. The workload is actually the pods that we deployed above.")
+		framework.DeployWorkloadOnControlplaneNode(ctx, framework.DeployWorkloadOnControlplaneNodeInput{
+			WorkloadCluster:                    workloadClient,
+			ControlPlane:                       controlplane,
+			WaitForDeploymentAvailableInterval: input.E2EConfig.GetIntervals(specName, "wait-deployment-available"),
+		})
+
+		By("Scale down the controlplane of the workload cluster and make sure that nodes running workload can be deleted even the draining process is blocked.")
+		// When we scale down the KCP, controlplane machines are by default deleted one by one, so it requires more time.
+		nodeDrainTimeoutKCPInterval := convertKCPDurationToInternal(controlplane.Spec.NodeDrainTimeout, controlPlaneReplicas)
+		framework.ScaleAndWaitControlPlane(ctx, framework.ScaleAndWaitControlPlaneInput{
+			ClusterProxy:        input.BootstrapClusterProxy,
+			Cluster:             cluster,
+			ControlPlane:        controlplane,
+			Replicas:            1,
+			WaitForControlPlane: nodeDrainTimeoutKCPInterval,
+		})
+
+		By("PASSED!")
+	})
+
+	AfterEach(func() {
+		// Dumps all the resources in the spec namespace, then cleanups the cluster object and the spec namespace itself.
+		dumpSpecResourcesAndCleanup(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, cancelWatches, cluster, input.E2EConfig.GetIntervals, input.SkipCleanup)
+	})
+}
+
+func convertDurationToInterval(duration *metav1.Duration, delayRate int) []interface{} {
+	minIntervalDuration := duration.Duration
+	maxIntervalDuration := (duration.Duration + time.Minute*2) * time.Duration(delayRate)
+	intervals := make([]interface{}, 0, 2)
+	intervals = append(intervals, maxIntervalDuration.String(), minIntervalDuration.String())
+	return intervals
+}
```

Comment:
I'm a little bit confused by this method, because AFAIK the intervals are passed to Eventually which expects a timeout and a polling interval, while here you are computing two intervals and adding 0,2 const.

Reply:
I misunderstand the polling interval. I will fix it. 
The reason for 2 minutes added to the interval because this timeout interval is calculated based on the NodeDrainTimeout. However, after NodeDrainTimeout is over, the cluster still needs more time to completely remove machines. That is why we need extra time for this process.